src/bundler.ts

```diff
@@ -147,33 +183,69 @@ class Bundler {
     }
     inlined.add(resolved);
     const backingDocument: ScannedDocument =
-        await analyzer._analyzeResolved(resolved);
+        await analyzer._scanResolved(resolved);
     const documentAst = dom5.parseFragment(backingDocument.document.contents);
     this.pathResolver.resolvePaths(documentAst, resolved, documentUrl);
-    const importParent = htmlImport.parentNode;
+    let importParent: ASTNode;
+    if (matchers.afterHiddenDiv(htmlImport)) {
+      importParent = dom5.nodeWalkPrior(htmlImport, matchers.hiddenDiv);
+      dom5.remove(htmlImport);
+      dom5.append(importParent, htmlImport);
+    } else if (matchers.beforeHiddenDiv(htmlImport)) {
+      const index = htmlImport.parentNode.childNodes.indexOf(htmlImport);
+      importParent = htmlImport.parentNode.childNodes[index + 1];
+      dom5.remove(htmlImport);
+      ASTUtils.prepend(importParent, htmlImport);
+    } else if (!matchers.inHiddenDiv(htmlImport)) {
+      let hiddenDiv = this.getHiddenNode();
+      dom5.replace(htmlImport, hiddenDiv);
+      dom5.append(hiddenDiv, htmlImport);
+      importParent = hiddenDiv;
+    } else {
+      importParent = htmlImport.parentNode;
+    }
+
+    ASTUtils.insertAllBefore(importParent, htmlImport, documentAst.childNodes);
     dom5.remove(htmlImport);
-    ASTUtils.prependMultiple(importParent, documentAst.childNodes);
   }
 
   async bundle(url: string): Promise<ASTNode> {
     const analyzer: Analyzer = new Analyzer(this.opts);
-    const analyzed: Document = await analyzer.analyzeRoot(url);
-    const newDocument = dom5.parse(analyzed.parsedDocument.contents);
+    const analyzedRoot: Document = await analyzer.analyzeRoot(url);
+    const newDocument = dom5.parse(analyzedRoot.parsedDocument.contents);
+
+    // Create a hidden div to target.
     const body = dom5.query(newDocument, matchers.body);
-    const head = dom5.query(newDocument, matchers.head);
-    const getNextImport = () => dom5.query(newDocument, matchers.htmlImport);
+    const hiddenDiv = this.getHiddenNode();
+
+    const getNextHtmlImport = () =>
+        dom5.query(newDocument, matchers.htmlImport);
     const elementInHead = dom5.predicates.parentMatches(matchers.head);
-    const inlinedImports = new Set<string>();
-    let c = 1;
-    for (let nextImport; nextImport = getNextImport();) {
+    const inlinedHtmlImports = new Set<string>();
+
+    let nextHtmlImport;
+    while (nextHtmlImport = getNextHtmlImport()) {
```

Comment:
It seems like there are potentially multiple walks over the document tree. Once here, and once for script inlining. Plus there's a re-parser of the contents. I wonder if this would be better as a walk over the original AST, cloning nodes to build up a new document on the fly...